code/Path_Distance_TableGenerator.py

```diff
@@ -0,0 +1,311 @@
+    cursor = conn.cursor()
+
+    # cursor.execute('UPDATE phys_nodes AS pnc \
+    # SET state = cp.state_province \
+    # FROM city_points AS cp \
+    # WHERE ABS(pnc.latitude - cp.city_latitude) < 0.001\
+    # AND ABS(pnc.longitude - cp.city_longitude) < 0.001 \
+    # AND pnc.city = cp.city_name \
+    # AND pnc.country = cp.country_code;')
+
+    # conn.commit()
+    # conn.close()
+
+    # Query the nodes and connections from the database
+    cursor.execute('SELECT node_name, city, state, country FROM phys_nodes')
+    nodes = cursor.fetchall()
+
+    cursor.execute('SELECT from_node, to_node FROM phys_nodes_conn')
+    connections = cursor.fetchall()
+
+    # Create a directed graph using networkx
+    G = nx.DiGraph()
+
+    # Add nodes to the graph
+    for node in nodes:
+        G.add_node(node[0], city=node[1], state=node[2], country=node[3])
+
+    # Add edges (connections) to the graph
+    for connection in connections:
+        if not connection[0] not in G.nodes and connection[1] not in G.nodes:
+            continue
+        G.add_edge(connection[0], connection[1])
+        # assume symmetric connections
+        G.add_edge(connection[1], connection[0])
+
+    # Find nodes matching the start and end city/state pairs
+    start_nodes = [node[0] for node in nodes if node[1] == start_city and
+                   (node[2] == start_state or not start_state) and
+                   node[3] == start_country]
+    end_nodes = [node[0] for node in nodes if node[1] == end_city and
+                 (node[2] == end_state or not end_state) and
+                 node[3] == end_country]
+
+    print('# of start nodes:', len(start_nodes), file=sys.stderr)
+    print('# of end nodes:', len(end_nodes), file=sys.stderr)
+
+    def _node_str(n):
+        return f'{n["city"]}/{n["state"]}/{n["country"]}'
+
+    # Find the shortest path for each combination of start and end nodes
+    all_paths = []
+    for start_node in start_nodes:
+        for end_node in end_nodes:
+            try:
+                shortest_path = nx.shortest_path(G, source=start_node, target=end_node)
+                shortest_path_location = [_node_str(G.nodes[node]) for node in shortest_path]
+                all_paths.append(shortest_path_location)
+            except NetworkXNoPath:
+                pass
+                #print(f"No path found from {start_city}, {start_country} to {end_city}, {end_country}")
+
+    # Print the result
+    #print(f"Shortest paths from {start_city}, {start_country} to {end_city}, {end_country}:")
+    path_distribution = defaultdict(int)
+    for path in all_paths:
+        path_distribution[str(path)] += 1
+    
+    table_rows = []
+
+    print(f"Path distribution from {start_city}, {start_state}, {start_country} to {end_city},{end_state}, {end_country}:", file=sys.stderr)
+    for path, count in path_distribution.items():
+        path = transform_location_format(path)
+        path = ast.literal_eval(path)
```

Comment:
Can you explain why do you need ast.literal_eval? What's the format of path before and after this call?

Reply:
as we get a path in the string format "[city1, city2, city3 .....]". we need to use ast.literal_eval to transform it into a list and calculate distance between pairs in it.